File: analyzer.py
"""
Single-pass analysis: reads the PCAP file once with filter 'sip or rtp'.

Compared to two-pass mode, this halves disk I/O and keeps CPU busy continuously
(no idle time between passes). Correct for all normal VoIP captures because SDP
always precedes RTP in a SIP signaling flow.

Limitation: RTP packets that arrive before their SDP has been seen (e.g. capture
started mid-call) are silently skipped — same behaviour as two-pass mode.
"""
from pathlib import Path
from typing import Dict, Tuple

from models import Call
from pcap_reader import combined_stream
from progress import Progress
from rtp_extractor import _PcapFileWriter, process_rtp_pkt
from sip_extractor import _SipFileCache, process_sip_pkt


def analyze(pcap_file: str, output_dir: Path) -> Dict[str, Call]:
    """
    Stream every SIP and RTP packet in a single tshark pass and write all output files.

    Returns the calls dict (call_id -> Call) for index.csv generation.
    """
    output_dir.mkdir(parents=True, exist_ok=True)

    calls: Dict[str, Call] = {}
    endpoint_map: Dict[Tuple[str, int], Call] = {}
    used_dirnames: set = set()
    pcap_writers: Dict[str, _PcapFileWriter] = {}
    wav_writers: Dict[str, object] = {}
    sip_cache = _SipFileCache()
    prog = Progress()

    try:
        for item in combined_stream(pcap_file):
            if item[0] == 'sip':
                _, sip_pkt = item
                process_sip_pkt(sip_pkt, calls, endpoint_map, used_dirnames, output_dir, sip_cache)
            else:
                _, ts, src_ip, src_port, dst_ip, dst_port, raw, udp_payload, datalink = item
                process_rtp_pkt(ts, src_ip, src_port, dst_ip, dst_port, raw, udp_payload, datalink,
                                endpoint_map, pcap_writers, wav_writers)
            prog.tick(calls=len(calls), streams=len(pcap_writers))
    finally:
        sip_cache.close_all()
        for w in pcap_writers.values():
            w.close()
        for w in wav_writers.values():
            w.close()

    prog.done(calls=len(calls), streams=len(pcap_writers))
    return calls

File: analyzer.go
package main

import (
	"fmt"
	"io"
	"os"
	"path/filepath"
)

// analyze performs a single-pass analysis over one or more PCAP files,
// processing both SIP and RTP packets while maintaining a shared call context.
// methodFilter: set of uppercase SIP methods to accept (empty = accept all).
// Returns the map of calls (call_id -> *Call).
func analyze(pcapFiles []string, outputDir string, methodFilter map[string]struct{}) (map[string]*Call, error) {
	if err := os.MkdirAll(outputDir, 0755); err != nil {
		return nil, fmt.Errorf("creating output directory: %w", err)
	}

	var totalBytes int64
	for _, f := range pcapFiles {
		if fi, err := os.Stat(f); err == nil {
			totalBytes += fi.Size()
		}
	}

	calls := make(map[string]*Call)
	endpointMap := make(map[Endpoint]*Call)
	usedDirs := make(map[string]struct{})
	rejectedIDs := make(map[string]struct{})
	pcapWriters := make(map[rtpKey]*PcapWriter)
	wavWriters := make(map[rtpKey]audioWriter)
	cache := newSipFileCache()
	prog := newProgress(totalBytes)

	defer func() {
		cache.closeAll()
		for _, pw := range pcapWriters {
			pw.Close()
		}
		for _, aw := range wavWriters {
			aw.close()
		}
	}()

	var bytesOffset int64
	for i, pcapFile := range pcapFiles {
		if len(pcapFiles) > 1 {
			logEvent("[file]  %s (%d/%d)", filepath.Base(pcapFile), i+1, len(pcapFiles))
		}

		reader, err := Open(pcapFile)
		if err != nil {
			return nil, fmt.Errorf("%s: %w", pcapFile, err)
		}
		datalink := reader.Datalink()

		for {
			ts, data, err := reader.Next()
			if err == io.EOF {
				break
			}
			if err != nil {
				reader.Close()
				return nil, fmt.Errorf("%s: reading packet: %w", pcapFile, err)
			}

			srcIP, dstIP, srcPort, dstPort, udpPayload := parseUDP(data, datalink)
			extra := fmt.Sprintf("%d calls  %d streams", len(calls), len(pcapWriters))
			if len(udpPayload) == 0 {
				prog.tick(bytesOffset+reader.BytesRead(), extra)
				continue
			}

			if (udpPayload[0] & 0xC0) == 0x80 {
				// RTP v2 packet
				if err := processRTPPkt(
					ts, srcIP, dstIP, srcPort, dstPort,
					data, udpPayload, datalink,
					endpointMap, pcapWriters, wavWriters, outputDir,
				); err != nil {
					logEvent("[warn]  %v", err)
				}
			} else if udpPayload[0] >= 0x20 && udpPayload[0] <= 0x7e {
				// Printable ASCII start — likely SIP
				pkt := sipPkt{
					timestamp: ts,
					srcIP:     srcIP,
					dstIP:     dstIP,
					srcPort:   srcPort,
					dstPort:   dstPort,
					payload:   udpPayload,
				}
				if err := processSIPPkt(pkt, calls, endpointMap, usedDirs, outputDir, cache, methodFilter, rejectedIDs); err != nil {
					logEvent("[warn]  %v", err)
				}
			}

			prog.tick(bytesOffset+reader.BytesRead(), extra)
		}

		bytesOffset += reader.BytesRead()
		reader.Close()
	}

	prog.done(totalBytes, fmt.Sprintf("%d calls  %d streams", len(calls), len(pcapWriters)))
	return calls, nil
}
